src/test/obj_ulog_advanced/obj_ulog_advanced.c

```diff
@@ -0,0 +1,408 @@
+ *
+ * Moreover, the DAOS developers observed a real issue occurring in their BMEM
+ * allocator which is based on PMEMOBJ (daos-stack/daos#11593). The issue
+ * occurred when the entries fell above the LANE_REDO_EXTERNAL_SIZE offset but
+ * there were not enough entries to trigger the persistent shadow log's growth
+ * (<= ULOG_BASE_SIZE - CACHELINE_SIZE = 960B).
+ *
+ * This test aims at reproducing this scenario to ensure this issue is not
+ * present in PMEMOBJ.
+ */
+
+/* exit code of a process when terminated in consequence of SIGABRT */
+#define SIGABRT_EXITSTATUS 134
+
+/* a single redo log entry's size - true only for a set-operation */
+#define ENTRY_SIZE (sizeof(struct ulog_entry_val))
+/*
+ * A 'slot' for the sake of this test means a single 64b value in persistent
+ * memory which set-operations will target.
+ *
+ * The number of slots is limited by the number of set-operations that can fit
+ * in a persistent shadow log before triggering its growth.
+ */
+#define SLOTS_NUM_MAX ((ULOG_BASE_SIZE - CACHELINE_SIZE) / ENTRY_SIZE) /* 60 */
+/* The number of set-operations that can fit in a single persistent redo log. */
+#define SLOTS_PER_REDO_LOG (LANE_REDO_EXTERNAL_SIZE / ENTRY_SIZE) /* 40 */
+
+struct root {
+	uint64_t slots[SLOTS_NUM_MAX];
+};
+
+/*
+ * It has to be big enough so the call counter won't reach this value naturally.
+ */
+#define BIG_ENOUGH_MAGIC_CALL_NUMBER 127
+
+/*
+ * The error injection is done for ulog_store().
+ * The abort injection is done for ulog_process().
+ *
+ * Both of them are run one after another not only in case of processing
+ * the user-built persistent shadow log but also whenever a reservation is
+ * necessary e.g. when additional persistent redo log is needed to accomodate
+ * the persistent shadow log. The persistent redo log reservation is done before
+ * processing the persistent shadow log hence -1.
+ *
+ * ERROR_INJECT_CALL() and ABORTED_CALL() prime the respective call counter to
+ * hit the dedicated magic value and trigger either an error injection or
+ * an abort injection.
+ */
+#define _ERROR_INJECT_CALL BIG_ENOUGH_MAGIC_CALL_NUMBER
+#define ERROR_INJECT_CALL(slots_num) \
+	((slots_num > SLOTS_PER_REDO_LOG) ? \
+		(_ERROR_INJECT_CALL - 1) : _ERROR_INJECT_CALL)
+
+#define _ABORTED_CALL (BIG_ENOUGH_MAGIC_CALL_NUMBER * 2)
+#define ABORTED_CALL(slots_num) \
+	((slots_num > SLOTS_PER_REDO_LOG) ? \
+		(_ABORTED_CALL - 1) : _ABORTED_CALL)
+
+FUNC_MOCK(ulog_store, void, struct ulog *dest, struct ulog *src, size_t nbytes,
+	size_t ulog_base_nbytes, size_t ulog_total_capacity,
+	struct ulog_next *next, const struct pmem_ops *p_ops)
+	/*
+	 * In this case, the error injection means replacing the number of bytes
+	 * truly populated in the persistent shadow log with the value reduced
+	 * to the capacity of a single persistent redo log. The error scenario
+	 * as envisioned by the issue which inspired the creation of this test.
+	 */
+	FUNC_MOCK_RUN(_ERROR_INJECT_CALL) {
```

Comment:
I found this confusing as a test :) Essentially the implementation is correct, but this is forcing an incorrect behavior. Essentially testing whether the test is correct. I know you have a big comment, but I had to spend some time thinking what this is for even though I know what's the problem.